burn-autodiff/src/ops/tensor.rs

```diff
@@ -672,6 +672,56 @@ impl<B: Backend> TensorOps<ADBackendDecorator<B>> for ADBackendDecorator<B> {
         }
     }
 
+    fn mask_scatter<const D: usize>(
+        tensor: ADTensor<B, D>,
+        mask: BoolTensor<B, D>,
+        source: ADTensor<B, D>,
+    ) -> ADTensor<B, D> {
+        #[derive(Debug)]
+        struct MaskScatter;
+
+        impl<B: Backend, const D: usize> Backward<B, D, 2> for MaskScatter {
+            type State = (BoolTensor<B, D>, Shape<D>, Shape<D>, B::Device);
+
+            fn backward(self, ops: Ops<Self::State, 2>, grads: &mut Gradients) {
+                let (mask, shape_lhs, shape_rhs, device) = ops.state;
+                let [mask_4lhs, mask_4rhs] = duplicate(&ops.parents, Some(mask));
+
+                binary::<B, D, D, D, _, _>(
+                    ops.parents,
+                    ops.node,
+                    grads,
+                    |grad| {
+                        let zeros = B::zeros(shape_lhs, &device);
+                        B::mask_scatter(grad, mask_4lhs.unwrap(), zeros)
+                    },
+                    |grad| {
+                        let zeros = B::zeros(shape_rhs, &device);
+                        B::mask_scatter(zeros, mask_4rhs.unwrap(), grad)
```

Comment:
I don't believe this is correct, I think the gradient of the LHS should be the gradient of the input tensor with a mask scatter using the same mask as input, but filled with the gradients of the RHS instead of zeros. And it's the reverse for the RHS.

Reply:
In term of memory management, everything seems just fine :) _